DynamicLinking.md

```diff
@@ -1,49 +1,102 @@
+dynamic loaders.
+
+Since the manner in which modules are loaded and instantiated is defined by the
+host environment (e.g., the [JavaScript API](JS.md)), dynamic linking requires
+use of host-specific functionality to link two modules. At a minimum, the host
+environment must provide a way to dynamically instantiate modules while
+connecting exports to imports.
+
+The simplest load-time dynamic linking scheme between modules A and B can be 
+achieved by having module A export functions, tables and memories that are
+imported by B. A C++ toolchain can expose this functionality by using the
+same function attributes currently used to export/import symbols from
+native DSOs/DLLs:
+```
+#ifdef _WIN32
+#  define EXPORT __declspec(dllexport)
+#  define IMPORT __declspec(dllimport)
+#else
+#  define EXPORT __attribute__ ((visibility ("default")))
+#  define IMPORT __attribute__ ((visibility ("default")))
+#endif
+
+typedef void (*PF)();
+
+IMPORT PF imp();
+EXPORT void exp() { imp()(); }
+```
+This code would generate a WebAssembly module with imports for:
+* the function `imp`
+* the table(s) used for indirectly-callable functions
+* the linear memory used to implement the heap
+
+and exports for:
+* the function `exp`
+* the imported table(s) and linear memory (so that other modules may
+  import this one)
+
+The host-specific loader can then simply connect the imports
+and exports at instantantiation-time.
+
+One extra detail is what
+to use as the [module name](Modules.md#imports) for imports (since
+WebAssembly has a two-level namespace). One option is to have a single default
+module name for all C/C++ imports/exports (which then allows the toolchain to
+put implementation-internal names in a separate module, avoiding the need for
+`__`-prefix conventions).
+
+To implement run-time dynamic linking (e.g., `dlopen` and `dlsym`):
+* `dlopen` would compile and instantiate a new module, storing the compiled
+  instance in a host-environment table, returning the index to the caller.
+* `dlsym` would be given this index, pull the instance out of the table,
+  search the instances's exports, append the found function to the function
+  table (using host-defined functionality in the MVP, but directly from
+  WebAssembly code in the
+  [future](FutureFeatures.md#more-table-operators-and-types)) and return the
+  table index of the appended element to the caller.
+
+Note that the representation of a C function-pointer in WebAssembly is an index
+into a function table, so the above scheme lines up perfectly with the
+function-pointer return value of `dlsym`.
+
+More complicated dynamic linking functionality (e.g., interposition, weak
+symbols, etc) can be simulated efficiently by assigning a function table
+index to each weak/mutable symbol, calling the symbol via `call_table` on that
+index, and mutating the underlying element as needed.
+
+After the MVP, we would like to standardize a single [ABI][] per source
```

Comment:
I think there might be multiple ABIs per source language, and they might not be standardized, at least not formally. I think something like "it may be useful in the future to standardize ABIs". That phrasing also removes "we" so it's more general.